docs/csharp/ref-locals-and-returns.md

```diff
@@ -0,0 +1,205 @@
+be advantageous to create a constant and return it by reference. But, if you
+are returning a reference to internal storage, you may want to enforce that
+the caller cannot modify the referenced storage. The following code
+defines a read-only property that returns a `readonly ref` to a `Point3D`
+that specifies the origin.
+
+[!code-csharp[OriginReference](../../samples/csharp/reference-semantics/Point3D.cs#OriginReference "Creating a readonly Origin reference")]
+
+When you assign a variable to a `ref return`, you can specify either a `ref readonly`
+variable, or a by-value copy of the readonly reference:
+
+[!code-csharp[AssignRefReadonly](../../samples/csharp/reference-semantics/Program.cs#AssignRefReadonly "Assigning a ref readonly")]
+
+The first assignment in the preceding code makes a copy of the `Origin` constant, and assigns
+that copy. The second assigns a reference. Notice that the `readonly` modifier
+must be part of the declaration of the variable. The reference to which it refers
+cannot be modified. Attempts to do so results in a compile-time error.
+
+## `readonly struct` type
+
+Applying `ref readonly` to high-traffic uses of a struct may be sufficient.
+Other times, you may want to create an immutable struct. Then, you can
+always pass by readonly reference. That practice 
+removes the defensive copies
+that take place when you access methods of a struct used as an `in` parameter.
+
+You can do that by creating a `readonly struct` type. You can add the `readonly`
+modifier to a struct declaration. The language enforces that all members of
+the struct are `readonly`: the `struct` must be immutable.
+
+There are other optimizations for a `readonly struct`. You can
+use the `in` modifier at every location where a `readonly struct` is an
+argument. In addition, you can return a `readonly struct` as a `ref return` when
+you are returning an object whose lifetime extends beyond the scope of the method
+returning the object.
+
+Finally, the compiler generates more efficient code when you call members of a
+`readonly struct`: The `this` reference is always an `in` parameter, instead
+of a copy of the receiver, passed by value to the member method. This optimization
+saves more copying when you use a `readonly struct`. The `Point3D` would be a great
+candidate for this change. The following code shows an update `ReadonlyPoint3D` structure:
+
+[!code-csharp[ReadonlyOnlyPoint3D](../../samples/csharp/reference-semantics/Point3D.cs#ReadonlyOnlyPoint3D "Defining an immutable structure")]
+
+## `ref struct` type
+
+Another related language feature is the ability to declare a value type that
+must be stack allocated. In other words, these types can never be created on the
+heap as a member of another class. The primary motivation for this feature
+was `Span<T>` and related structures, but you can create `ref struct` types
+when you have similar requirements. In this article, you see examples
```

Comment:
The requirements aren't clear. Can you mention a use case here?

Reply:
I re-wrote that section. Likely, using `Span<T>` and related are the most common use case. I tried to make that point, and point to other uses.